### PR TITLE
Add middleware

### PR DESCRIPTION
**Pull request by novellat**
## Hey, I just made a Pull Request!

When migrating to a new backend system, the old backend sometimes uses middleware for its plugins. This change allows to migrate plugins with middleware.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by novellat** [comment]
@benjaminr @tudi2d hey guys. I'm sorry to bother I just wonder what do you think of these changes. 

It seems that many projects have used middleware and this change was quite helpful to us. What are your thuoghts - should it go to the mainline? Or are there any tweaks I can make?

**Comment by benjdlambert** [comment]
Hey @novellat, sorry for the delay in this one. 

Had a busy few weeks with KubeCon and BackstageCon and some PTO.

Hmm, I think whats a little confusing for me is the usage of the `auth` middleware in the docs. Ideally with the release of > 1.24 we want to avoid using that middleware at all in the new backend system, as it's supported out of the box by default.

I'm wondering if with that said, is this change still useful? We'd recommend using the framework instead of that `auth` middleware at all I think.

**Comment by novellat** [comment]
@benjdlambert 
Hey Ben, absolutely no worries, hope you had a good time :) 

When it comes to auth middleware, I understand your point. In our case, we're using multiple pieces of middleware (we even have a logout middleware!) and since we haven't migrated to using the framework, I used it as an example.

Of course, we're planning to upgrade to the newest practices, but the first step was to switfly upgrade to the new backend system. And having the optional middleware argument allows for this.

So the question to you is: shall we close this PR, or shall I just updatethe docs? (eg, changing from createAuthMiddleware to, let's say, createCustomMiddleware, or any other suggestion you might have).

File: docs/backend-system/building-backends/08-migrating.md [hunk 1]
```diff
@@ -48,8 +48,9 @@ async function main() {
   // ... early init
   const createEnv = makeCreateEnv(config);
   const todoEnv = useHotMemoize(module, () => createEnv('todo')); // repeated for N plugins
+  const authMiddleware = await createAuthMiddleware(todoEnv); // optional middleware
   const apiRouter = Router();
-  apiRouter.use('/todo', await todo(todoEnv)); // repeated for N plugins
+  apiRouter.use('/todo', authMiddleware, await todo(todoEnv)); // repeated for N plugins
```
Comment (vinzscam) [inline docs/backend-system/building-backends/08-migrating.md:53]: shouldn't this snippet stay as it was as having `authMiddleware` isn't common?

File: docs/backend-system/building-backends/08-migrating.md [hunk 2]
```diff
@@ -98,7 +99,9 @@ import { legacyPlugin } from '@backstage/backend-common';
 
 const backend = createBackend();
 /* highlight-add-next-line */
-backend.add(legacyPlugin('todo', import('./plugins/todo')));
+backend.add(
```
Comment (vinzscam) [inline docs/backend-system/building-backends/08-migrating.md:102]: mmm you can already do this without changing the signature of the `.add` method.

`backend.add(legacyPlugin('todo', Router().use(yourMiddleware, await todo(todoEnv))))`

I don't think changing the `.add` method is the right way to go, as this is just pure express logic.

**Review by vinzscam — CHANGES_REQUESTED** [review]

File: docs/backend-system/building-backends/08-migrating.md (hunk 2 above)
Comment (novellat) [reply docs/backend-system/building-backends/08-migrating.md:102]: hmm, that's fair. I didn't want to expose router, but this is probably the easiest way (albeit not documented/mentioned)

**Pull request closed without merging**